src/cacheline.1
.\"                                      -*- nroff -*-
.\" cacheline.1 - Manual page for cacheline.
.\"
.\" Copyright (C) 2001-2005 Oskar Liljeblad
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU Library General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH CACHELINE "1" "May 4, 2005" "cacheline (regex-markup)"
.SH NAME
cacheline \- Delay newlines in output
.SH SYNOPSIS
.B cacheline
.SH DESCRIPTION
This manual page document describes the \fBcacheline\fP command.

The cacheline program reads from standard in and writes to standard
out. Whenever a newline character is read, it is inhibited and not
printed until another character is read. This way, when monitoring
files on a terminal (e.g. with tail), the bottom row (line) can be
used. (Normally this line is always empty due to the last newline
character printed.)
.SH OPTIONS
None.
.SH EXAMPLES
Monitor a syslog file:
.br
	\fBtail -f /var/log/syslog | cacheline\fB
.SH REPORTING BUGS
Report bugs to <\fIoskar@osk.mine.nu\fP>.
.SH AUTHOR
\fBcacheline\fP was written by Oskar Liljeblad <\fIoskar@osk.mine.nu\fP>.
.SH COPYRIGHT
Copyright \(co 2001-2005 Oskar Liljeblad

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


NEWS
2005-05-22: regex-markup 0.10.0 released.
  Localization support added.
  Use gnulib for improved compatibility.
  Only use lstat64 when available (reported by Michal Fojtik).
  Tested on FreeBSD 4.10 (thanks to Michal Fojtik).
  Update manual pages, add copyright and bug reporting address.
  Cleaned up RPM package spec file.
  Added full GNU GPL header to source files and manual pages.
  Support --help and --version in cacheline.
  Lexer code now includes config.h before everything else.

2003-04-11: regex-markup 0.9.0 released.
  Implemented new command 'set', to simplify substitution constructs.
  Upgraded to autoconf 2.57 and automake 1.7.3.
  Fixed bug that would cause remark to die with "cannot allocate memory"
  when using $n where n is out of range in substitution rules.
  Added new testsuites bug05, e82, e83, e84, e85, e86.
  Clarified README about rule format location.
  Never ignore ctrl+c/SIGINT - I don't remember the benefit of ignoring it
  in the first place. If someone knows, please mail me. :)

2002-01-06: regex-markup 0.8.0 released.
  gcc 2.95 is now required to compile regex-markup.
  Fixed a bug with /(..)*/ matches that would cause regex-markup to
  highlight incorrectly or die ("Cannot allocate memory").
  Fixed a bug in s/../../ rules that would produce wrong output.
  Added example rules for make.
  Improved syslog highlighting rules.

2001-12-21: regex-markup 0.7.0 released.
  Added section about common problems to the README file.
  Remark now also looks in $HOME/.remark and /usr/share/regex-markup for
  rule files.
  The example rule files are installed in /usr/share/regex-markup by
  default.
  Added files for making an RPM package of regex-markup.

2001-12-19: regex-markup 0.6.0 released.
  Cleaned up code to reduce unnecessary memory allocations.
  Added new example rules for diff.
  Rewrote the cacheline program has in C.
  Fixed how escaped characters are dealt with in substitution strings.
  Improved terminal width detection in logwatcher.

2001-12-17: regex-markup 0.5.0 released.
  Almost all code has been rewritten and improved, resulting in more
  features, faster execution, cleaner and less buggy code.
  Implemented line wrapping in remark (previously this was done
  separately). As a result of this, foldlog was removed and logwatcher
  updated to use the build-in wrapping.
  Added a bunch of test cases (to a total of 558 tests).

2001-12-03: regex-markup 0.4.0 released.
  Began support for Cygwin by adding compatibility files from fileutils.
  Updated copyright.
  Added a new program "foldlog" to fold syslog lines.
  Added a testsuite for foldlog.
  Added missing file bison.simple.fixed (needed for complete rebuild).
  Improved error messages.
  Improved the syslog highlighter (logwatcher).
  Improved memory management (don't allocate and free as much memory).
  Fixed a bug in the parsing code that would cause freed memory to be read.

2001-10-28: regex-markup 0.3.0 released.
  Updated manual page.
  Fixed a substitution bug that would cause segmentation violation.
  Fixed style bugs causing incorrect highlighting.
  Better signal handling.
  Improved substitution with support for $`, $&, and $' variables.

2001-10-25: regex-markup 0.2.0 released.
  Added manual pages for remark.
  Added files for making a regex-markup Debian package.
  The author has changed email-address to oskar@osk.mine.nu.

2001-08-13: regex-markup 0.1.0 released.
  First public release.


src/cacheline.c
/* cacheline.c - Delays printing of last newline
 *
 * Copyright (C) 2001-2005 Oskar Liljeblad
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <config.h>
#include <getopt.h>		/* Gnulib/GNU Libc/POSIX */
#include <stdbool.h>		/* Gnulib/C99/POSIX */
#include <stdlib.h>		/* C89 */
#include <stdio.h>		/* C89 */
#include "gettext.h"		/* Gnulib */
#define _(String) gettext(String)
#include "getline.h"		/* Gnulib/GNU Libc */
#include "version-etc.h"	/* Gnulib */

enum {
    HELP_OPT = 256,
    VERSION_OPT
};

static struct option long_opts[] = {
    { "help", no_argument, NULL, HELP_OPT },
    { "version", no_argument, NULL, VERSION_OPT },
    { 0, 0, 0, 0 }
};

const char version_etc_copyright[] = "Copyright (C) 2001-2005 Oskar Liljeblad";

int
main(int argc, char **argv)
{
    char *buf = NULL;
    size_t buflen = 0;
    bool keep = false;

    setlocale(LC_ALL, "");
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);

    while (true) {
        int c;

        c = getopt_long(argc, argv, "", long_opts, NULL);
        if (c == -1)
            break;

        switch (c) {
        case HELP_OPT:
            printf(_("Usage: %s [OPTION]...\n"), argv[0]);
            puts(_("Read from in and write to standard out, delaying newline until a complete\n"
                 "line has been read.\n"));
            printf(_("      --help     display this help and exit\n"));
            printf(_("      --version  output version information and exit\n"));
            printf(_("\nReport bugs to <%s>.\n"), PACKAGE_BUGREPORT);
            exit(EXIT_SUCCESS);
        case VERSION_OPT:
            version_etc(stdout, "cacheline", PACKAGE, VERSION, "Oskar Liljeblad", NULL);
            exit(EXIT_SUCCESS);
        }
    }

    for (;;) {
        ssize_t linelen;

        linelen = getline(&buf, &buflen, stdin);
        if (linelen < 0) {
	    if (ferror(stdin)) {
	    	perror(argv[0]);
	    	exit(1);
	    }
	    break;
	}

	if (linelen > 0) {
	    if (keep) {
	    	fputc('\n', stdout);
		keep = false;
	    }
	    if (buf[linelen-1] == '\n') {
	    	keep = true;
	    	linelen--;
	    }
	    fwrite(buf, linelen, 1, stdout);
	    fflush(stdout);
	}
    }
    
    exit(0);
}
